pkg/storage/stores/shipper/storage/client.go

```diff
@@ -0,0 +1,94 @@
+package storage
+
+import (
+	"context"
+	"io"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/grafana/loki/pkg/storage/chunk"
+)
+
+const delimiter = "/"
+
+type Client interface {
+	ListTables(ctx context.Context) ([]string, error)
+	ListFiles(ctx context.Context, tableName string) ([]IndexFile, error)
+	GetFile(ctx context.Context, tableName, fileName string) (io.ReadCloser, error)
+	PutFile(ctx context.Context, tableName, fileName string, file io.ReadSeeker) error
+	DeleteFile(ctx context.Context, tableName, fileName string) error
+	IsObjectNotFoundErr(err error) bool
+	Stop()
+}
+
+type indexStorageClient struct {
+	objectClient  chunk.ObjectClient
+	storagePrefix string
+}
+
+type IndexFile struct {
+	Name       string
+	ModifiedAt time.Time
+}
+
+func NewIndexStorageClient(objectClient chunk.ObjectClient, storagePrefix string) Client {
+	return &indexStorageClient{objectClient: objectClient, storagePrefix: storagePrefix}
+}
+
+func (s *indexStorageClient) ListTables(ctx context.Context) ([]string, error) {
+	_, tables, err := s.objectClient.List(ctx, s.storagePrefix, delimiter)
+	if err != nil {
+		return nil, err
+	}
+
+	tableNames := make([]string, 0, len(tables))
+	for _, table := range tables {
+		tableNames = append(tableNames, path.Base(string(table)))
+	}
+
+	return tableNames, nil
+}
+
+func (s *indexStorageClient) ListFiles(ctx context.Context, tableName string) ([]IndexFile, error) {
+	// The forward slash here needs to stay because we are trying to list contents of a directory without which we will get the name of the same directory back with hosted object stores.
```

Comment:
tiny nit: it helps to keep these comment lines a bit shorter so we don't have to scroll :)